crates/red_knot_python_semantic/src/types/infer.rs

```diff
@@ -2471,7 +2471,48 @@ impl<'db> TypeInferenceBuilder<'db> {
             (UnaryOp::USub, Type::BooleanLiteral(bool)) => Type::IntLiteral(-i64::from(bool)),
             (UnaryOp::Invert, Type::BooleanLiteral(bool)) => Type::IntLiteral(!i64::from(bool)),
 
-            (UnaryOp::Not, ty) => ty.bool(self.db).negate().into_type(self.db),
+            (UnaryOp::Not, ty) => {
+                // TODO: move to Type::bool method
+                // I initially wrote this code in the bool method but it caused a cycle. And since
+                // the cycle detection was not turned on for that method it was crashing.
+                let bool_type = if let Type::Instance(class) = ty {
```

Comment:
Moving this code to https://github.com/astral-sh/ruff/pull/13827/files#diff-42314c006689490bbdfbeeb973de64046b3e069e3d88f67520aeba375f20e655R757 will cause stack overflow.